File: src/derivation/conceptual-query.ts
export interface QueryPathStep {
  from: string
  predicate: string
  to: string
  inverse: boolean
}

export interface ConceptualQueryResult {
  path: QueryPathStep[]
  filters: Array<{ field: string; value: string }>
  rootNoun?: string
}

export interface Reading {
  text: string
  nouns: string[] // nouns[0] is the "subject" noun, nouns[1+] are object nouns
}

/**
 * Extract quoted literal values (e.g. 'High') from the query text.
 * Returns the stripped query and the extracted values in order.
 */
function extractFilters(query: string): {
  stripped: string
  values: string[]
} {
  const values: string[] = []
  const stripped = query.replace(/'([^']+)'/g, (_match, val: string) => {
    values.push(val)
    return ''
  })
  return { stripped: stripped.replace(/\s+/g, ' ').trim(), values }
}

/**
 * Find all known nouns that appear in a text segment.
 * Matches longest noun names first to avoid partial matches
 * (e.g. "Support Request" before "Support").
 */
function findNounsInSegment(segment: string, nouns: string[]): string[] {
  // Sort by length descending for longest-first matching
  const sorted = [...nouns].sort((a, b) => b.length - a.length)
  const found: string[] = []
  let remaining = segment

  for (const noun of sorted) {
    // Case-insensitive search for the noun in the remaining text
    const idx = remaining.toLowerCase().indexOf(noun.toLowerCase())
    if (idx !== -1) {
      found.push(noun)
      // Remove the matched noun to prevent overlapping matches
      remaining =
        remaining.slice(0, idx) + remaining.slice(idx + noun.length)
    }
  }

  return found
}

/**
 * Find a reading that connects two nouns, checking both forward and inverse directions.
 *
 * Forward: `from` is nouns[0] in the reading (the subject position).
 * Inverse: `from` appears in the reading but is NOT nouns[0].
 */
function findReading(
  from: string,
  to: string,
  readings: Reading[],
): { reading: Reading; inverse: boolean } | undefined {
  // Forward: from is the first noun in the reading
  const forward = readings.find(
    (r) =>
      r.nouns.length >= 2 &&
      r.nouns[0].toLowerCase() === from.toLowerCase() &&
      r.nouns.some((n) => n.toLowerCase() === to.toLowerCase()),
  )
  if (forward) return { reading: forward, inverse: false }

  // Inverse: from appears in the reading but is not the first noun
  const inverse = readings.find(
    (r) =>
      r.nouns.length >= 2 &&
      r.nouns.some((n) => n.toLowerCase() === from.toLowerCase()) &&
      r.nouns.some((n) => n.toLowerCase() === to.toLowerCase()) &&
      r.nouns[0].toLowerCase() !== from.toLowerCase(),
  )
  if (inverse) return { reading: inverse, inverse: true }

  return undefined
}

/**
 * Extract the predicate text from a reading given the two nouns involved.
 * For "Customer submits Support Request", the predicate is "submits".
 */
function extractPredicate(reading: Reading): string {
  let text = reading.text
  // Remove nouns from the reading text to isolate the predicate(s).
  // Sort nouns by length descending to remove longest first.
  const sorted = [...reading.nouns].sort((a, b) => b.length - a.length)
  for (const noun of sorted) {
    text = text.replace(new RegExp(escapeRegex(noun), 'gi'), '')
  }
  return text.replace(/\s+/g, ' ').trim()
}

function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

/**
 * Resolve a conceptual ORM query into a structured path of reading hops with filters.
 *
 * Algorithm:
 * 1. Extract quoted literal values ('High') as filters
 * 2. Split query on " that " to get segments
 * 3. Find nouns in each segment (longest-first matching)
 * 4. For each consecutive pair of nouns, find a reading connecting them (forward or inverse)
 * 5. Build path of QueryPathSteps
 * 6. Associate filters with the target noun of the segment where the filter value appeared
 */
export function resolveConceptualQuery(
  query: string,
  nouns: string[],
  readings: Reading[],
): ConceptualQueryResult {
  const { stripped, values: filterValues } = extractFilters(query)

  // Split on " that " to get segments
  const segments = stripped.split(/\s+that\s+/i)

  // Collect all nouns found across segments, in order
  const nounSequence: string[] = []

  // For each segment, find the nouns it contains
  const segmentNouns: string[][] = segments.map((seg) =>
    findNounsInSegment(seg, nouns),
  )

  // Build the ordered noun sequence from segment nouns.
  // The first segment contributes all its nouns.
  // Subsequent segments contribute nouns that aren't already the last noun in sequence.
  for (let i = 0; i < segmentNouns.length; i++) {
    for (const noun of segmentNouns[i]) {
      if (
        nounSequence.length === 0 ||
        nounSequence[nounSequence.length - 1].toLowerCase() !==
          noun.toLowerCase()
      ) {
        nounSequence.push(noun)
      }
    }
  }

  // If we found fewer than 1 known noun, return empty result
  if (nounSequence.length === 0) {
    return { path: [], filters: [] }
  }

  if (nounSequence.length === 1) {
    return { path: [], filters: [], rootNoun: nounSequence[0] }
  }

  // Build path steps from the noun sequence.
  // For each target noun, try the previous noun first; if no reading exists,
  // walk backward through earlier nouns to find one that connects.
  // This handles fan-out queries like "X that has A that has B" where both
  // A and B connect to X rather than chaining A -> B.
  const path: QueryPathStep[] = []
  const rootNoun = nounSequence[0]

  for (let i = 1; i < nounSequence.length; i++) {
    const to = nounSequence[i]
    let matched = false

    // Try connecting from the immediately preceding noun first, then walk back
    for (let j = i - 1; j >= 0; j--) {
      const from = nounSequence[j]
      const match = findReading(from, to, readings)
      if (match) {
        const predicate = extractPredicate(match.reading)
        path.push({
          from,
          predicate,
          to,
          inverse: match.inverse,
        })
        matched = true
        break
      }
    }

    if (!matched) {
      // Can't find any reading connecting to this noun — skip it
      continue
    }
  }

  // If we couldn't build any path, and the nouns weren't recognized via readings,
  // return empty
  if (path.length === 0) {
    return { path: [], filters: [] }
  }

  // Map filter values to the target noun of the path step they belong to.
  // Strategy: walk through segments that have filter values (identified by
  // the original query having a quoted value in that segment), and associate
  // the filter with the last noun in that segment.
  const filters: Array<{ field: string; value: string }> = []

  // Re-split the original query on " that " to find which segments have filter values
  const originalSegments = query.split(/\s+that\s+/i)
  let filterIdx = 0
  for (const seg of originalSegments) {
    // Count quoted values in this segment
    const quotedMatches = seg.match(/'([^']+)'/g)
    if (quotedMatches) {
      // Find the last noun in this segment to use as the field
      const segNouns = findNounsInSegment(
        seg.replace(/'[^']+'/g, ''),
        nouns,
      )
      const fieldNoun = segNouns[segNouns.length - 1]

      for (const _qm of quotedMatches) {
        if (filterIdx < filterValues.length && fieldNoun) {
          filters.push({ field: fieldNoun, value: filterValues[filterIdx] })
        }
        filterIdx++
      }
    }
  }

  return { path, filters, rootNoun }
}

File: crates/fol-engine/src/conceptual_query.rs
// crates/fol-engine/src/conceptual_query.rs
//
// Conceptual query resolution — natural language query → path of reading hops.
//
// "Support Requests that have Priority 'High'" resolves to:
//   path: [{ from: "SupportRequest", predicate: "has", to: "Priority", inverse: false }]
//   filters: [{ field: "Priority", value: "High" }]
//
// Algorithm:
// 1. Extract quoted literals as filters
// 2. Split on "that" to get segments
// 3. Find nouns in each segment (longest-first matching)
// 4. For each pair, find a reading connecting them
// 5. Build path + associate filters with target nouns

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryPathStep {
    pub from: String,
    pub predicate: String,
    pub to: String,
    pub inverse: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConceptualQueryResult {
    pub path: Vec<QueryPathStep>,
    pub filters: Vec<Filter>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub root_noun: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Filter {
    pub field: String,
    pub value: String,
}

pub struct Reading {
    pub text: String,
    pub nouns: Vec<String>,
}

fn extract_filters(query: &str) -> (String, Vec<String>) {
    let mut values = Vec::new();
    let mut stripped = String::new();
    let mut in_quote = false;
    let mut current_val = String::new();

    for ch in query.chars() {
        if ch == '\'' {
            if in_quote {
                values.push(current_val.clone());
                current_val.clear();
            }
            in_quote = !in_quote;
        } else if in_quote {
            current_val.push(ch);
        } else {
            stripped.push(ch);
        }
    }

    // Normalize whitespace
    let stripped = stripped.split_whitespace().collect::<Vec<_>>().join(" ");
    (stripped, values)
}

fn find_nouns_in_segment(segment: &str, nouns: &[String]) -> Vec<String> {
    let mut sorted: Vec<&String> = nouns.iter().collect();
    sorted.sort_by(|a, b| b.len().cmp(&a.len()));

    let mut found = Vec::new();
    let mut remaining = segment.to_lowercase();

    for noun in sorted {
        let lower = noun.to_lowercase();
        if remaining.contains(&lower) {
            found.push(noun.clone());
            remaining = remaining.replacen(&lower, "", 1);
        }
    }

    found
}

fn find_reading<'a>(from: &str, to: &str, readings: &'a [Reading]) -> Option<(&'a Reading, bool)> {
    let from_lower = from.to_lowercase();
    let to_lower = to.to_lowercase();

    // Forward: from is first noun
    let forward = readings.iter().find(|r| {
        r.nouns.len() >= 2
            && r.nouns[0].to_lowercase() == from_lower
            && r.nouns.iter().any(|n| n.to_lowercase() == to_lower)
    });
    if let Some(r) = forward { return Some((r, false)) }

    // Inverse: from appears but is not first
    let inverse = readings.iter().find(|r| {
        r.nouns.len() >= 2
            && r.nouns.iter().any(|n| n.to_lowercase() == from_lower)
            && r.nouns.iter().any(|n| n.to_lowercase() == to_lower)
            && r.nouns[0].to_lowercase() != from_lower
    });
    if let Some(r) = inverse { return Some((r, true)) }

    None
}

fn extract_predicate(reading: &Reading) -> String {
    let mut text = reading.text.clone();
    let mut sorted = reading.nouns.clone();
    sorted.sort_by(|a, b| b.len().cmp(&a.len()));
    for noun in &sorted {
        text = text.replace(noun, "");
    }
    text.split_whitespace().collect::<Vec<_>>().join(" ")
}

pub fn resolve_conceptual_query(
    query: &str,
    nouns: &[String],
    readings: &[Reading],
) -> ConceptualQueryResult {
    let (stripped, filter_values) = extract_filters(query);

    let segments: Vec<&str> = stripped.split(" that ").collect();
    let segment_nouns: Vec<Vec<String>> = segments.iter()
        .map(|seg| find_nouns_in_segment(seg, nouns))
        .collect();

    let mut noun_sequence = Vec::new();
    for seg_nouns in &segment_nouns {
        for noun in seg_nouns {
            if noun_sequence.is_empty() || noun_sequence.last().map(|n: &String| n.to_lowercase()) != Some(noun.to_lowercase()) {
                noun_sequence.push(noun.clone());
            }
        }
    }

    if noun_sequence.is_empty() {
        return ConceptualQueryResult { path: vec![], filters: vec![], root_noun: None };
    }
    if noun_sequence.len() == 1 {
        return ConceptualQueryResult { path: vec![], filters: vec![], root_noun: Some(noun_sequence[0].clone()) };
    }

    let mut path = Vec::new();
    let root_noun = noun_sequence[0].clone();

    for i in 1..noun_sequence.len() {
        let to = &noun_sequence[i];
        for j in (0..i).rev() {
            let from = &noun_sequence[j];
            if let Some((reading, inverse)) = find_reading(from, to, readings) {
                let predicate = extract_predicate(reading);
                path.push(QueryPathStep {
                    from: from.clone(),
                    predicate,
                    to: to.clone(),
                    inverse,
                });
                break;
            }
        }
    }

    if path.is_empty() {
        return ConceptualQueryResult { path: vec![], filters: vec![], root_noun: None };
    }

    // Map filters to target nouns
    let mut filters = Vec::new();
    let original_segments: Vec<&str> = query.split(" that ").collect();
    let mut filter_idx = 0;
    for seg in &original_segments {
        let count = seg.matches('\'').count() / 2;
        if count > 0 {
            let seg_nouns = find_nouns_in_segment(&seg.replace('\'', ""), nouns);
            let field = seg_nouns.last().cloned().unwrap_or_default();
            for _ in 0..count {
                if filter_idx < filter_values.len() && !field.is_empty() {
                    filters.push(Filter { field: field.clone(), value: filter_values[filter_idx].clone() });
                }
                filter_idx += 1;
            }
        }
    }

    ConceptualQueryResult { path, filters, root_noun: Some(root_noun) }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_binary_query() {
        let nouns = vec!["Person".to_string(), "Name".to_string()];
        let readings = vec![Reading { text: "Person has Name".to_string(), nouns: vec!["Person".to_string(), "Name".to_string()] }];
        let result = resolve_conceptual_query("Person that has Name", &nouns, &readings);
        assert_eq!(result.path.len(), 1);
        assert_eq!(result.path[0].from, "Person");
        assert_eq!(result.path[0].to, "Name");
        assert!(!result.path[0].inverse);
    }

    #[test]
    fn query_with_filter() {
        let nouns = vec!["SupportRequest".to_string(), "Priority".to_string()];
        let readings = vec![Reading { text: "SupportRequest has Priority".to_string(), nouns: vec!["SupportRequest".to_string(), "Priority".to_string()] }];
        let result = resolve_conceptual_query("SupportRequest that has Priority 'High'", &nouns, &readings);
        assert_eq!(result.path.len(), 1);
        assert_eq!(result.filters.len(), 1);
        assert_eq!(result.filters[0].field, "Priority");
        assert_eq!(result.filters[0].value, "High");
    }

    #[test]
    fn single_noun_returns_root() {
        let nouns = vec!["Person".to_string()];
        let result = resolve_conceptual_query("Person", &nouns, &[]);
        assert_eq!(result.path.len(), 0);
        assert_eq!(result.root_noun.as_deref(), Some("Person"));
    }

    #[test]
    fn inverse_reading() {
        let nouns = vec!["Department".to_string(), "Academic".to_string()];
        let readings = vec![Reading { text: "Academic works for Department".to_string(), nouns: vec!["Academic".to_string(), "Department".to_string()] }];
        let result = resolve_conceptual_query("Department that Academic works for", &nouns, &readings);
        assert_eq!(result.path.len(), 1);
        assert!(result.path[0].inverse);
    }
}
